### PR TITLE
Touch ID Integration

### PR DESCRIPTION
**Pull request by 7hommay**
- Use Touch ID instead of PIN when the device supports it and has Touch ID enabled
- Prompts the user to upgrade an account to use Touch ID after a successful login
- To prevent code duplication, most of the enrollment & authentication logic is moved to the new ChallengeService

### CODE REVIEW

File: Tiqr/Classes/AuthenticationConfirmViewController.m [hunk 1]
```diff
@@ -43,15 +51,17 @@ @interface AuthenticationConfirmViewController ()
...
 @end
 
 @implementation AuthenticationConfirmViewController
 
-- (instancetype)initWithAuthenticationChallenge:(AuthenticationChallenge *)challenge {
+- (instancetype)init {
     self = [super initWithNibName:@"AuthenticationConfirmView" bundle:nil];
 	if (self != nil) {
-		self.challenge = challenge;
+		self.challenge = ServiceContainer.sharedInstance.challengeService.currentAuthenticationChallenge;
```
Comment (ijansch) [inline Tiqr/Classes/AuthenticationConfirmViewController.m]: Why move from injection based (initWith) to state based (current... on the challengeService)? Isn't injection based more loosely coupled?

File: Tiqr/Classes/AuthenticationSummaryViewController.m [hunk 2]
```diff
@@ -44,15 +45,17 @@ @interface AuthenticationSummaryViewController ()
...
 
 @implementation AuthenticationSummaryViewController
 
-- (instancetype)initWithAuthenticationChallenge:(AuthenticationChallenge *)challenge {
+- (instancetype)initWithUsedPIN:(NSString *)PIN {
     self = [super initWithNibName:@"AuthenticationSummaryView" bundle:nil];
 	if (self != nil) {
-		self.challenge = challenge;
+		self.challenge = ServiceContainer.sharedInstance.challengeService.currentAuthenticationChallenge;
+        self.PIN = PIN;
```
Comment (ijansch) [inline Tiqr/Classes/AuthenticationSummaryViewController.m:58]: Putting the PIN in a member might mean it gets kept around in memory longer than needed, potentially making it vulnerable to an attacker looking at the app's memory. Is this avoidable?

File: Tiqr/Classes/ChallengeService.h [hunk 3]
```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2010-2011 SURFnet bv
```
Comment (ijansch) [inline Tiqr/Classes/ChallengeService.h]: If it's a new file update the year plz.

File: Tiqr/Classes/AuthenticationSummaryViewController.m (hunk 2 above)
Comment (7hommay) [inline Tiqr/Classes/AuthenticationSummaryViewController.m:58]: Maybe this is avoidable by always performing the upgrade to TouchID after a successful sign in, while keeping the original stored data in the keychain. When the user has decided if he wants to upgrade or not, the original data or the upgraded data can be deleted

**Pull request merged**